### PR TITLE
feat: add java modules

### PR DESCRIPTION
**Pull request by jeromevdl**
*Issue #, if available:*

*Description of changes:* Adding modules (lambda functions) for java runtime

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by dreamorosi** [comment]
After merging #14 I have added the conditional logic to deploy the modules in each language by running `LANGUAGE=nodejs npm run infra:deploy` (see contributing doc for possible values).

Please rebase `main` and incorporate your proposed functions in the `if/else` statement for the appropriate language.

**Comment by scottgerring** [comment]
Sure! I'll have feedback for you by tomorrow.

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import software.amazon.lambda.powertools.logging.LoggingUtils;
+import software.amazon.lambda.powertools.metrics.Metrics;
+import software.amazon.lambda.powertools.tracing.Tracing;
+
+import static com.amazonaws.powertools.workshop.Utils.getImageMetadata;
+import static com.amazonaws.powertools.workshop.Utils.markFileAs;
+import static software.amazon.lambda.powertools.metrics.MetricsUtils.metricsLogger;
+
+/**
+ * Handler for requests to Lambda function.
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java:33]: Perhaps some more text about what it does?

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        LoggingUtils.appendKey("S3Object", object.toString());
+
+        try {
+            // Mark file as working
+            markFileAs(object.getFileId(), "in-progress", null);
+
+            String newObjectKey = processOneIdempotently(object);
+
+            // Mark file as done
+            markFileAs(object.getFileId(), "completed", newObjectKey);
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java:85]: If this fails, we'll flip back to `markFileAs(..., "in-progress", ...)` and then hit the idemp cache on :82 . is that fine? Could we also do all of this in the idempotent method?

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+
+        return newObjectKey;
+    }
+
+    private byte[] createThumbnail(byte[] originalImageBytes) throws IOException {
+        BufferedImage bufferedImage = Thumbnails.of(new ByteArrayInputStream(originalImageBytes))
+                .size(TRANSFORMED_IMAGE_WIDTH, TRANSFORMED_IMAGE_HEIGHT)
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java:127]: assert that original image size >= resized image size?

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/S3EBEvent.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+    public void setResources(List<String> resources) {
+        this.resources = resources;
+    }
+
+    public static class S3EBEventDetail {
+        private String version;
+        private Map<String, String> bucket;
+        private Map<String, String> object;
+        @JsonProperty("request-id")
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/S3EBEvent.java:98]: This is extremely nitpicky, but I find these hard to read when there are annotations and no intermediate newlines  🥲

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    public static final DynamoDbClient ddbClient = DynamoDbClient.builder()
+            .region(AWS_REGION)
+            .credentialsProvider(credentialsProvider)
+            .build();
+
+    public static Map<String, String> getImageMetadata(String key) {
+        String fileId = extractFileId(key);
+        GetItemResponse response = ddbClient.getItem(GetItemRequest
+                .builder()
+                .tableName(TABLE_NAME_FILES)
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:59]: Feels a bit janky to pull configuration in a utils class from an environment variable

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+                .key(Map.of("id", AttributeValue.builder().s(fileId).build()))
+                .attributesToGet("id", "userId")
+                .build());
+
+        if (!response.hasItem()) {
+            throw new RuntimeException("File metadata not found");
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:65]: Is this exceptional, or a null return?

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                .build();
+
+        // Perform AWS Signature Version 4 signing
+        Aws4Signer signer = Aws4Signer.create();
+        SdkHttpFullRequest prepRequest = signer.sign(request, attributes);
+
+        // Perform the HTTP call to GraphQL endpoint
+        HttpPost httpPost = new HttpPost(endpoint);
+        for (Map.Entry<String, List<String>> header : prepRequest.headers().entrySet()) {
+            if (header.getKey().equalsIgnoreCase("host")) { continue; }
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:158]: @jeromevdl nitpick - I feel like you've discouraged me from using `continue;` before; you could also just invert this expression and put the block below inside

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        } catch (IOException e) {
+            throw new UpdateFileStatusException(e);
+        }
+    }
+
+    private static String performHttpCall(CloseableHttpClient client, HttpPost httpPost) {
+        try (var response = client.execute(httpPost)) {
+            String responseData = new String(response.getEntity().getContent().readAllBytes(), StandardCharsets.UTF_8);
+            if (responseData.contains("errors")) {
+                throw new UpdateFileStatusException(responseData);
```
Comment (scottgerring) [inline functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:176]: The generic name of the method `performHttpCall` janks with it being tightly coupled to updating file statuses

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.logging.log4j.Logger;
+import software.amazon.lambda.powertools.batch.BatchMessageHandlerBuilder;
+import software.amazon.lambda.powertools.batch.handler.BatchMessageHandler;
+import software.amazon.lambda.powertools.logging.Logging;
+import software.amazon.lambda.powertools.logging.LoggingUtils;
+import software.amazon.lambda.powertools.metrics.Metrics;
+import software.amazon.lambda.powertools.tracing.Tracing;
+
+/**
+ * Handler for requests to Lambda function.
```
Comment (scottgerring) [inline functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java:22]: What's it do?

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+        String userId = data.get("userId").getS();
+        String transformedFileKey = data.get("transformedFileKey").getS();
+
+        // Add the file id and user id to the logger so that all the logs after this
+        // will have these attributes and we can correlate them
+        LoggingUtils.appendKey("fileId", fileId);
+        LoggingUtils.appendKey("userId", userId);
+        try {
+            // Get the labels from Rekognition
+            getLabels(BUCKET_NAME_FILES, fileId, userId, transformedFileKey);
```
Comment (scottgerring) [inline functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java:61]: Could be idempotent to save recalculating labels?

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+        List<Label> labels = detectLabelsResponse.labels();
+        if (labels == null || labels.isEmpty()) {
+            throw new NoLabelsFoundException(ImageMetadata.of(fileId, userId));
+        }
+
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info(labels.stream().map(label -> format("%s (%.2f)", label.name(), label.confidence())).collect(Collectors.joining(",", "labels=[", "]")));
+        }
+
+        boolean personLabel = labels.stream().anyMatch(label -> label.name().contains("Person") && label.confidence() != null && label.confidence() > 75);
```
Comment (scottgerring) [inline functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java:81]: Java needs a null coalescing operator ....

File: functions/java/modules/module3/src/main/java/com/amazonaws/powertools/workshop/Module3Handler.java [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import com.amazonaws.services.lambda.runtime.RequestHandler;
+import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;
+import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyResponseEvent;
+import java.util.Map;
+import software.amazon.lambda.powertools.logging.Logging;
+import software.amazon.lambda.powertools.metrics.Metrics;
+import software.amazon.lambda.powertools.tracing.Tracing;
+
+/**
+ * Handler for requests to Lambda function.
```
Comment (scottgerring) [inline functions/java/modules/module3/src/main/java/com/amazonaws/powertools/workshop/Module3Handler.java:13]: What's it do?

**Review by scottgerring — CHANGES_REQUESTED** [review]
Some fairly pedantic feedback inline @jeromevdl :)

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java (hunk 2 above)
Comment (jeromevdl) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java:85]: this is how they do in typescript, I keep the logic

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java (hunk 3 above)
Comment (jeromevdl) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Module1Handler.java:127]: we don't really care...

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 5 above)
Comment (jeromevdl) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:59]: It's not the most beautiful code I wrote, I kept the same "structure" as typescript... If you want to refactor feel free, I have no more time to do this.

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 6 above)
Comment (jeromevdl) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:65]: nope that's an error

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 7 above)
Comment (jeromevdl) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:158]: thank you but you're nitpicking. It's a not powertools itself ;)

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java (hunk 10 above)
Comment (jeromevdl) [reply functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Module2Handler.java:61]: not the objective of module 2, not in typescript.

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 11 above)
Comment (jeromevdl) [reply functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java:81]: ??

File: functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 7 above)
Comment (scottgerring) [reply functions/java/modules/module1/src/main/java/com/amazonaws/powertools/workshop/Utils.java:158]: I have only one code review mode 🤣  It's up to you to decide what makes sense for the scope. It is _very_ nitpicky.

File: functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java (hunk 11 above)
Comment (scottgerring) [reply functions/java/modules/module2/src/main/java/com/amazonaws/powertools/workshop/Utils.java:81]:
```
bob?.bill?.fred?.toString()
```

vs

```
if (bob != null && bob.bill != null && bob.bill.fred != null) {
   return fred.toString()
else {
   return null;
}

In this case it doesn't make much difference - there's only one level of nulls - I am just being mean to java.

**Review by scottgerring — APPROVED** [review]

**Comment by dreamorosi** [comment]
Thank you both for the help with this!

**Pull request merged**